### PR TITLE
change `pub enum` to `ENUM!` or `FLAGS!`

### PR DESCRIPTION
**Pull request by Eh2406**
I often removed the numbers ware redundant

### CODE REVIEW

**Comment by Eh2406** [comment]
notes for further work on `pub enum`:

many enum's have `#[allow(unused_qualifications)]` but I did not find docs on that allow. Do you want me to convert throws as well?

sapi SpeechAudioFormatType and vss VSS_APPLICATION_LEVEL have an enum val of -1. How should I convert that to a u32?

shobjidl has DEFAULT as a val in two different enum's this is not compatible with `ENUM!`

**Comment by retep998** [comment]
The dummy variants for univariant enums can be removed when converting them to `ENUM!` and `FLAGS!`.
I'd prefer if the explicit discriminants weren't removed.
`#[allow(unused_qualifications)]` isn't needed when switching to `ENUM!`/`FLAGS!`.
For negative discriminants do `-1i32 as u32`.

**Comment by retep998** [comment]
For those enums in shobjidl, it turns out that many enums in winapi had their variant names shortened to remove the common prefix. Like for example.
```C
enum FDE_OVERWRITE_RESPONSE
    {
        FDEOR_DEFAULT	= 0,
        FDEOR_ACCEPT	= 1,
        FDEOR_REFUSE	= 2
    } 	FDE_OVERWRITE_RESPONSE;
```
It is currently represented in winapi as
```Rust
#[repr(i32)] #[derive(Clone, Copy, Debug)]
pub enum FDE_OVERWRITE_RESPONSE {
    DEFAULT = 0,
    ACCEPT = 1,
    REFUSE = 2,
}
```
What it should become:
```Rust
ENUM!{enum FDE_OVERWRITE_RESPONSE {
    FDEOR_DEFAULT = 0,
    FDEOR_ACCEPT = 1,
    FDEOR_REFUSE = 2,
}}

**Comment by Eh2406** [comment]
I will continue working on this when I get a chance. Thanks for all the work you are putting into this project!

**Comment by Eh2406** [comment]
reverted the numbers. Next commit will do the `#[allow(unused_qualifications)]`

**Comment by Eh2406** [comment]
ok so I converted all of the `pub enum` to `ENUM!` or `FLAGS!` but then started getting cryptic error messages. I reverted and added one file at a time to isolate the issues. The ones that can be added are in my repo, if they meet with your approval, I will squash.

Then I went back into files to figure out exactly which change is at fault.
For example D3DSHADER_INSTRUCTION_OPCODE_TYPE hits the macro recursion limit.
Tracking these down is slow. the error messages just go to the macros file. Perhaps they are for a spret pr?

**Comment by retep998** [comment]
For larger enums, the solution is to specify all the discriminants explicitly, otherwise it has to use macro recursion in order to figure them out. Oh how I wish Rust had stable procedural macros.

**Comment by Eh2406** [comment]
That is all of them! Please review, so I can squash.

File: src/corsym.rs [hunk 1]
```diff
@@ -65,14 +60,12 @@ interface ISymUnmanagedBinder(ISymUnmanagedBinderVtbl): IUnknown(IUnknownVtbl) {
     ) -> ::HRESULT
 }
 );
-#[repr(i32)] #[derive(Clone, Copy, Debug)] #[allow(unused_qualifications)]
-pub enum CorSymSearchPolicyAttributes {
+ENUM!{enum CorSymSearchPolicyAttributes {
```
Comment (retep998) [inline src/corsym.rs]: This should probably be `FLAGS!`

File: src/d2d1.rs [hunk 2]
```diff
@@ -173,25 +145,19 @@ STRUCT!{struct D2D1_TRIANGLE {
     point2: ::D2D1_POINT_2F,
     point3: ::D2D1_POINT_2F,
 }}
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum D2D1_PATH_SEGMENT {
+ENUM!{enum D2D1_PATH_SEGMENT {
```
Comment (retep998) [inline src/d2d1.rs]: This should be `FLAGS!`

File: src/d2d1.rs [hunk 3]
```diff
@@ -237,40 +201,30 @@ STRUCT!{struct D2D1_LAYER_PARAMETERS {
...
+ENUM!{enum D2D1_RENDER_TARGET_USAGE {
     D2D1_RENDER_TARGET_USAGE_NONE = 0x00000000,
     D2D1_RENDER_TARGET_USAGE_FORCE_BITMAP_REMOTING = 0x00000001,
     D2D1_RENDER_TARGET_USAGE_GDI_COMPATIBLE = 0x00000002,
-}
-pub use self::D2D1_RENDER_TARGET_USAGE::*;
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum D2D1_PRESENT_OPTIONS {
+}}
+ENUM!{enum D2D1_PRESENT_OPTIONS {
```
Comment (retep998) [inline src/d2d1.rs]: Also `FLAGS!`.
A whole bunch of these enums here that have hexadecimal discriminants should be `FLAGS!`.

File: src/http.rs [hunk 4]
```diff
@@ -516,7 +501,8 @@ pub struct HTTP_REQUEST_HEADERS {
     pub pUnknownHeaders: PHTTP_UNKNOWN_HEADER,
     pub TrailerCount: ::USHORT,
     pub pTrailers: PHTTP_UNKNOWN_HEADER,
-    pub KnownHeaders: [HTTP_KNOWN_HEADER; HTTP_HEADER_ID::HttpHeaderRequestMaximum as usize],
+    pub KnownHeaders: [HTTP_KNOWN_HEADER; 41usize],
+    //                                    ^- FIXME: need to be HttpHeaderRequestMaximum
```
Comment (retep998) [inline src/http.rs]: Would doing `HttpHeaderRequestMaximum.0` work here?

File: src/processsnapshot.rs [hunk 5]
```diff
@@ -1,8 +1,7 @@
 // Copyright © 2015, skdltmxn
 // Licensed under the MIT License <LICENSE.md>
 //! Defines the process snapshot API
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum PSS_CAPTURE_FLAGS {
+ENUM!{enum PSS_CAPTURE_FLAGS {
```
Comment (retep998) [inline src/processsnapshot.rs]: `FLAGS!`

File: src/processsnapshot.rs [hunk 6]
```diff
@@ -34,19 +32,17 @@ pub enum PSS_QUERY_INFORMATION_CLASS {
...
+ENUM!{enum PSS_WALK_INFORMATION_CLASS {
     PSS_WALK_AUXILIARY_PAGES = 0,
     PSS_WALK_VA_SPACE = 1,
     PSS_WALK_HANDLES = 2,
     PSS_WALK_THREADS = 3,
-}
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum PSS_DUPLICATE_FLAGS {
+}}
+ENUM!{enum PSS_DUPLICATE_FLAGS {
```
Comment (retep998) [inline src/processsnapshot.rs]: `FLAGS!`

File: src/shobjidl.rs [hunk 7]
```diff
@@ -439,26 +439,24 @@ pub struct IContactManagerInterop;
...
-    DESKTOPABSOLUTEEDITING = 0x8004c000u32 as i32,
-    FILESYSPATH = 0x80058000u32 as i32,
-    URL = 0x80068000u32 as i32,
-    PARENTRELATIVEFORADDRESSBAR = 0x8007c001u32 as i32,
-    PARENTRELATIVE = 0x80080001u32 as i32,
-    PARENTRELATIVEFORUI = 0x80094001u32 as i32,
-}
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum SICHINTF {
+    PARENTRELATIVEPARSING = 0x80018001u32,
```
Comment (retep998) [inline src/shobjidl.rs]: No need to have the u32 suffix here anymore.

File: src/shobjidl.rs [hunk 8]
```diff
@@ -439,26 +439,24 @@ pub struct IContactManagerInterop;
 //4498
 pub type SFGAOF = ::ULONG;
 //9466
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum SIGDN {
+ENUM!{enum SIGDN {
     NORMALDISPLAY = 0,
```
Comment (retep998) [inline src/shobjidl.rs]: Variant names are missing their prefix.

File: src/shobjidl.rs [hunk 9]
```diff
@@ -439,26 +439,24 @@ pub struct IContactManagerInterop;
...
+    PARENTRELATIVEEDITING = 0x80031001u32,
+    DESKTOPABSOLUTEEDITING = 0x8004c000u32,
+    FILESYSPATH = 0x80058000u32,
+    URL = 0x80068000u32,
+    PARENTRELATIVEFORADDRESSBAR = 0x8007c001u32,
+    PARENTRELATIVE = 0x80080001u32,
+    PARENTRELATIVEFORUI = 0x80094001u32,
+}}
+ENUM!{enum SICHINTF {
     DISPLAY = 0,
```
Comment (retep998) [inline src/shobjidl.rs]: More missing prefixes.

File: src/shobjidl.rs [hunk 10]
```diff
@@ -515,8 +510,7 @@ interface IFileDialogEvents(IFileDialogEventsVtbl): IUnknown(IUnknownVtbl) {
     ) -> ::HRESULT
 }
 );
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum FILEOPENDIALOGOPTIONS {
+ENUM!{enum FILEOPENDIALOGOPTIONS {
```
Comment (retep998) [inline src/shobjidl.rs]: Missing prefixes and `FLAGS!`

File: src/shobjidl.rs [hunk 11]
```diff
@@ -589,13 +583,12 @@ interface IFileOpenDialog(IFileOpenDialogVtbl): IFileDialog(IFileDialogVtbl) {
     fn GetSelectedItems(&mut self, ppsai: *mut *mut IShellItemArray) -> ::HRESULT
 }
 );
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum CDCONTROLSTATEF {
+ENUM!{enum CDCONTROLSTATEF {
     INACTIVE = 0x00000000,
```
Comment (retep998) [inline src/shobjidl.rs]: Missing prefixes.

File: src/windowscodecs.rs [hunk 12]
```diff
@@ -11,35 +11,26 @@ STRUCT!{struct WICRect {
...
+ENUM!{enum WICBitmapEncoderCacheOption {
     WICBitmapEncoderCacheInMemory = 0,
     WICBitmapEncoderCacheTempFile = 0x1,
     WICBitmapEncoderNoCache = 0x2,
-}
-pub use self::WICBitmapEncoderCacheOption::*;
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum WICComponentType {
+}}
+ENUM!{enum WICComponentType {
```
Comment (retep998) [inline src/windowscodecs.rs]: `FLAGS!`

File: src/windowscodecs.rs [hunk 13]
```diff
@@ -47,18 +38,14 @@ pub enum WICComponentType {
     WICMetadataWriter = 0x10,
     WICPixelFormat = 0x20,
     WICAllComponents = 0x3f,
-}
-pub use self::WICComponentType::*;
-#[repr(i32)] #[derive(Copy, Clone, Debug)]
-#[allow(unused_qualifications, overflowing_literals, trivial_numeric_casts)]
-pub enum WICComponentEnumerateOptions {
+}}
+ENUM!{enum WICComponentEnumerateOptions {
```
Comment (retep998) [inline src/windowscodecs.rs]: `FLAGS!`

File: src/windowscodecs.rs [hunk 14]
```diff
@@ -105,68 +87,50 @@ pub enum WICBitmapDitherType {
...
+ENUM!{enum WICBitmapAlphaChannelOption {
     WICBitmapUseAlpha = 0,
     WICBitmapUsePremultipliedAlpha = 0x1,
     WICBitmapIgnoreAlpha = 0x2,
-}
-pub use self::WICBitmapAlphaChannelOption::*;
-#[repr(i32)] #[derive(Copy, Clone, Debug)] #[allow(unused_qualifications)]
-pub enum WICBitmapTransformOptions {
+}}
+ENUM!{enum WICBitmapTransformOptions {
```
Comment (retep998) [inline src/windowscodecs.rs]: `FLAGS!`

File: src/http.rs (hunk 4 above)
Comment (Eh2406) [inline src/http.rs]: no, I get compiler error.

**Comment by Eh2406** [comment]
squashed and ready.

when I next have time I perhaps I can work on
```
#[repr(C)] #[derive(Clone, Copy, Debug)]
pub struct IExecuteCommandApplicationHostEnvironment;
```
I.E. the many empty struct and turning them into `STRUCT!`

**Comment by retep998** [comment]
Actually all those `IFoo` empty structs are supposed to be `RIDL!` interfaces, it's just that nobody took the time to flesh them out yet.

File: src/commctrl.rs [hunk 15]
```diff
@@ -3433,8 +3431,7 @@ pub const CB_GETCUEBANNER: ::UINT = CBM_FIRST + 4;
 pub type PFTASKDIALOGCALLBACK = Option<unsafe extern "system" fn(
     hwnd: ::HWND, msg: ::UINT, wParam: ::WPARAM, lParam: ::LPARAM, lpRefData: ::LONG_PTR,
 ) -> ::HRESULT>;
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum TASKDIALOG_FLAGS {
+ENUM!{enum TASKDIALOG_FLAGS {
```
Comment (retep998) [inline src/commctrl.rs]: `FLAGS!`

File: src/shobjidl.rs [hunk 16]
```diff
@@ -515,8 +510,7 @@ interface IFileDialogEvents(IFileDialogEventsVtbl): IUnknown(IUnknownVtbl) {
     ) -> ::HRESULT
 }
 );
-#[repr(i32)] #[derive(Clone, Copy, Debug)]
-pub enum FILEOPENDIALOGOPTIONS {
+FLAGS!{enum FILEOPENDIALOGOPTIONS {
     OVERWRITEPROMPT = 0x2,
```
Comment (retep998) [inline src/shobjidl.rs]: Missing prefixes on variants

**Pull request merged**